python-api/livy/client.py

```diff
@@ -0,0 +1,444 @@
+        future : concurrent.futures.Future
+            A future to monitor the status of the job
+
+        Examples
+        -------
+        >>> file_handle = open("/test_package.egg")
+
+        >>> client.upload_pyfile(file_handle)
+
+        >>> def upload_pyfile_job(context):
+        >>>    # Importing module from test_package.egg
+        >>>    from test.pyfile_test import TestClass
+        >>>    test_class = TestClass()
+        >>>    return test_class.say_hello()
+
+        >>> client.submit(upload_pyfile_job)
+        """
+        return self._upload_file_or_pyfile("upload-pyfile", open_file)
+
+    def stop(self, shutdown_context):
+        """
+        Stops the remote context.
+        The function will return immediately and will not wait for the pending
+        jobs to get completed
+
+        Parameters
+        ----------
+        shutdown_context : Boolean
+            Whether to shutdown the underlying Spark context. If false, the
+            context will keep running and it's still possible to send commands
+            to it, if the backend being used supports it.
+        """
+        with self.lock:
+            if not self._stopped:
+                self._executor.shutdown(wait=False)
+                try:
+                    if shutdown_context:
+                        session_uri = "/" + str(self._session_id)
+                        headers = {'X-Requested-By': 'livy'}
+                        self._conn.send_request("DELETE", session_uri,
+                                                headers=headers)
+                except:
+                    raise Exception(traceback.format_exc())
+                self._stopped = True
+
+    def _set_uri(self, uri):
+        if uri is not None and uri.scheme in ('http', 'https'):
+            self._config.set(self._CONFIG_SECTION, 'livy.uri', uri.geturl())
+        else:
+            url_exception = uri.geturl if uri is not None else None
+            raise ValueError('Cannot create client - Uri not supported - ',
+                             url_exception)
+
+    def _set_conf(self, key, value):
+        if value is not None:
+            self._config.set(self._CONFIG_SECTION, key, value)
+        else:
+            self._delete_conf(key)
+
+    def _delete_conf(self, key):
+        self._config.remove_option(self._CONFIG_SECTION, key)
+
+    def _set_multiple_conf(self, conf_dict):
+        for key, value in conf_dict.iteritems():
+            self._set_conf(key, value)
+
+    def _load_config(self, load_defaults=True, conf_dict=None):
+        self._config.add_section(self._CONFIG_SECTION)
+        if load_defaults:
+            self._load_default_config()
+        if conf_dict is not None and len(conf_dict) > 0:
+            self._set_multiple_conf(conf_dict)
+
+    def _load_default_config(self):
+        config_dir = os.environ.get(self._LIVY_CLIENT_CONF_DIR)
+        if config_dir is None:
+            raise KeyError('Config directory not set in environment')
```

Comment:
This should not thrown an exception, because the default state of any user's machine is to not have this env variable set.